File: searchingAlgorithms/binarySearch/binary_Search.go
package main

import "fmt"
		

func main() {
	
//Implementing binarySearch

	arr := []int{0,2,4,5,6,8}

	var result int = binarySearch(arr,0,len(arr)-1,8) //arguments : array,smallest index,largest index,value to be searched

	if result==-1 {
		fmt.Println("Element not present")
	}else{
		fmt.Println("Element present at index: ",result);
	}
}


func binarySearch(arr []int,l int, r int, x int) int {

	if r>=l {
		var mid int = l+(r-l)/2 
			if arr[mid]==x {
				return mid
			}
			if arr[mid]<x {
				return binarySearch(arr, mid+1, r, x)
			}

		return binarySearch(arr, l, mid-1, x)
	}
	return -1	
}

File: searchingAlgorithms/binarySearch/BinarySearch.java
public class BinarySearch{

	public static int BinSearch(int[] arr, int value){

		int low = 0;
		int high = arr.length - 1;
		int mid;

		do{

			mid = low/2 + high/2;

			if(arr[mid] == value){

				return mid;

			}
			else if(value > arr[mid]){

				low = mid + 1;

			}
			else if(value < arr[mid]){

				high = mid;

			}

//			System.out.println("low: " + low + " high: " + high);

		}while(high - low > 0);

		if(arr[high] == value){

			return high;

		}

		return -1;

	}

	public static void main(String[] args){

		int[] arr = {3, 4, 6, 17, 20, 77};

		int index = BinSearch(arr, 6);

		System.out.println(index);

	}

}